subgraph-js/src/types.ts

```diff
@@ -62,6 +69,99 @@ function addTypeNameToPossibleReturn<T>(
+  maybeObject: PromiseOrValue<null | T>,
+  typename: string,
+): PromiseOrValue<null | (T & { __typename?: string })> {
+  if (isPromise(maybeObject)) {
+    return maybeObject.then((x: any) => addTypeNameToPossibleReturn(x, typename));
+  }
+  return addTypeNameToPossibleReturn(maybeObject, typename);
+}
+
+/**
+ * Copied from GraphQL-js.
+ *
+ * For @key on interfaces, we need to check that we can resolve the runtime type of the object returned by the interface
+ * `__resolveReference`. If we cannot, and we simply don't add any `__typename` to the result, the graphQL-js will end
+ * erroring out, but the error will not be user friendly as it will say something along the lines of:
+ * ```
+ * Abstract type "_Entity" must resolve to an Object type at runtime for field "Query._entities". Either the "_Entity" type
+ * should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.
+ * ```
+ * But this is ultimately incorrect, as it is only interface type the user must use "resolveType", add a __typename, or rely
+ * on "isTypeOf". And so we have to somewhat copy this. As we do, we also adapt the message a bit to be more user friendly.
+ */
+function ensureValidRuntimeType(
+  runtimeTypeName: unknown,
+  schema: GraphQLSchema,
+  returnType: GraphQLAbstractType,
+  result: unknown,
+): GraphQLObjectType {
+  if (runtimeTypeName == null) {
+    throw new GraphQLError(
+      `Abstract type "${returnType.name}" \`__resolveReference\` method must resolve to an Object type at runtime. Either the object returned by "${returnType}.__resolveReference" must include a valid \`__typename\` field, or the "${returnType.name}" type should provide a "resolveType" function or each possible type should provide an "isTypeOf" function.`,
+    );
+  }
+
+  if (typeof runtimeTypeName !== 'string') {
+    throw new GraphQLError(
+      `Abstract type "${returnType.name}" \`__resolveReference\` method must resolve to an Object type at runtime with ` +
+        `value ${inspect(result)}, received "${inspect(runtimeTypeName)}".`,
+    );
+  }
+
+  const runtimeType = schema.getType(runtimeTypeName);
+  if (runtimeType == null) {
+    throw new GraphQLError(
+      `Abstract type "${returnType.name}" \`__resolveReference\` method resolved to a type "${runtimeTypeName}" that does not exist inside the schema.`,
+    );
+  }
+
+  if (!isObjectType(runtimeType)) {
+    throw new GraphQLError(
+      `Abstract type "${returnType.name}" \`__resolveReference\` method resolved to a non-object type "${runtimeTypeName}".`,
+    );
+  }
+
+  if (!schema.isSubType(returnType, runtimeType)) {
+    throw new GraphQLError(
+      `Runtime Object type "${runtimeType.name}" \`__resolveReference\` method is not a possible type for "${returnType.name}".`,
+    );
+  }
+
+  return runtimeType;
+}
+
+function withResolvedType<T>({
+  type,
+  value,
+  context,
+  info,
+  callback,
+}: {
+  type: GraphQLInterfaceType,
+  value: any,
```

Comment:
I'm open to concrete suggestions but it's not obvious to me what we can do much better.

More precisely:
- `value` is the result of the call to the `__referenceResolver` method, and we do define as `any` [here](https://github.com/apollographql/federation/blob/main/subgraph-js/src/schemaExtensions.ts#L12), and as its ultimately written by end users, I'm not sure we want to add new assumptions (and it can genuinely kind of can be almost anything in practice).
- `context` is that `TContext` generic argument thing that `graphql-js` passes everywhere to resolvers, but there is no constraint on it in practice so it can be anything, and it's not like maintaining the generic type gives us any benefit here in our own code. I mean, I guess we can make this and plenty of other places be generic on `TContext`, though it'll add noise for not much benefits and that feel pretty unrelated to the problem of this PR in the first place.